File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int DiameterOfBinaryTree(TreeNode root) {

        int diameter = 0;
        dfs(root, ref diameter);
        return diameter;
           
    }

    public int dfs(TreeNode node, ref int diameter){
        if(node == null) return 0;

        int left = dfs(node.left, ref diameter);
        int right = dfs(node.right, ref diameter);

        diameter = Math.Max(diameter, left + right);

        return Math.Max(left, right) + 1;

    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        self.diameter = 0

        def getMaxDepth(node):
            if not node:
                return 0

            left = getMaxDepth(node.left)
            right = getMaxDepth(node.right)

            # Diameter is the longest path through this node
            self.diameter = max(self.diameter, left + right)

            # Return depth of the subtree
            return max(left, right) + 1

        getMaxDepth(root)
        return self.diameter

        
   



        
        